docs/migration-to-NRI.md

```diff
@@ -0,0 +1,227 @@
+# Prerequisities
+
+- Up and running CRI Resource Manager
+- One of the two supported policies in use: balloons or topology-aware.
+- For other policies a little bit more work is required and the policies need to be 'ported'. This can be done by just following the example of how balloons or topology-aware were converted.
+
+# Steps for an initial/basic migration test
+
+## Containerd
+
+Replace the containerd version in the system with 1.7 or newer version (NRI server not supported in older versions).
+
+Replace kubelet's --container-runtime-endpoint=/var/run/cri-resmgr/cri-resmgr.sock with --container-runtime-endpoint=/var/run/containerd/containerd.sock
+
+Replacing the runtime endpoint on a node that was setup using Kubeadm:
+```console 
+# Get the Kubelet args
+systemctl cat kubelet <- Look for: EnvironmentFile=/.../kubeadm-flags.env
+
+vim /.../kubeadm-flags.env
+  KUBELET_KUBEADM_ARGS="--container-runtime-endpoint=unix:///var/run/containerd/containerd.sock --pod-infra-container-image=registry.k8s.io/pause:3.9"
+
+vim /etc/sysconfig/kubelet 
+  KUBELET_EXTRA_ARGS= --container-runtime-endpoint=/var/run/containerd/containerd.sock <- Remember this aswell
+
+systemctl restart containerd
+systemctl restart kubelet
+```
+
+Edit the containerd config file and look for the section [plugins."io.containerd.nri.v1.nri"] and replace "disable = true" with "disable = false":
+```console
+vim /etc/containerd/config.toml
+```
+```toml
+[plugins."io.containerd.nri.v1.nri"]
+  disable = false
+  disable_connections = false
+  plugin_config_path = "/etc/nri/conf.d"
+  plugin_path = "/opt/nri/plugins"
+  plugin_registration_timeout = "5s"
+  plugin_request_timeout = "2s"
+  socket_path = "/var/run/nri/nri.sock"
+```
+
+## CRI-O
+
+Ensure that crio version 1.26.2 or newer is used.
+
+Replace kubelet's --container-runtime-endpoint=/var/run/cri-resmgr/cri-resmgr.sock with --container-runtime-endpoint=/var/run/crio/crio.sock
+
+Replacing the runtime endpoint on a node that was setup using Kubeadm:
+```console 
+# Get the Kubelet args
+systemctl cat kubelet <- Look for: EnvironmentFile=/.../kubeadm-flags.env
+
+vim /.../kubeadm-flags.env
+  KUBELET_KUBEADM_ARGS="--container-runtime-endpoint=unix:///var/run/crio/crio.sock --pod-infra-container-image=registry.k8s.io/pause:3.9"
+
+vim /etc/sysconfig/kubelet 
+  KUBELET_EXTRA_ARGS= --container-runtime-endpoint=/var/run/crio/crio.sock <- Remember this aswell
+
+systemctl restart crio
+systemctl restart kubelet
+```
+
+Edit the crio config and look for the "[crio.nri]" section, replace "enable_nri = false" with "enable_nri = true" and uncomment the NRI config lines:
+```console
+vim /etc/crio/crio.conf
+```
+
+```
+# CRI-O NRI configuration.
+[crio.nri]
+
+# Globally enable or disable NRI.
+enable_nri = true
+
+# NRI configuration file to use.
+nri_config_file = "/etc/nri/nri.conf"
+
+# NRI socket to listen on.
+nri_listen = "/var/run/nri/nri.sock"
+
+# NRI plugin directory to use.
+nri_plugin_dir = "/opt/nri/plugins"
```

Comment:
II think it would be better to either use crio itself to enable NRI (so it injects any defaults as comments as well)

```
CRIO_CONF=/etc/crio/crio.conf
cp $CRIO_CONF $CRIO_CONF.orig
crio --enable-nri config > $CRIO_CONF
```

or to use a drop-in fragment. I *think* something like this should work:

```
cat > /etc/crio/crio.conf.d/10-enable-nri.conf <<EOF
[crio.nri]
enable_nri = true
EOF
```

The problem with the above instructions is that they can easily break something in a future setup if the CRI-O NRI configuration change (especially any defaults in there).

Reply:
Shouldn't be a problem. The commented lines cri-o outputs for the `config` subcommand are its defaults. So you don't need to uncomment them unless you want to change the configuration to diverge from the defaults.

Reply:
But if you do uncomment them, then you are not running any more with the defaults. You are running with values which are explicitly set to the same as the current default values. If the defaults change, then in the former case you will still run with the updated defaults, but in the latter case you will keep running with your forced values which used to be the defaults but are not any more. So there is a difference and (usually) therefore if the defaults (at any point in time) are fine for you, you shouldn't uncomment them in the config.

Reply:
Thanks for clearing that up!

Reply:
> or to use a drop-in fragment. I _think_ something like this should work:
> 
> ```
> cat > /etc/crio/crio.conf.d/10-enable-nri.conf <<EOF
> [crio.nri]
> enable_nri = true
> EOF
> ```
> 
> The problem with the above instructions is that they can easily break something in a future setup if the CRI-O NRI configuration change (especially any defaults in there).

BTW, I used this variant in the nri-plugins repo e2e tests, and cri-o worked just fine with nri resource plugin.